File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/Examen19-20/Weapon.rb
#encoding: UTF-8

require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace

	class Weapon

		def initialize(n,t,u) # n es string, t es WeaponType, u es int
			@name = n
			@type = t
			@uses = u
		end
		
		def self.newCopy(other)
			new(other.name, other.type, other.uses)
		end
		
		def getUIversion
			WeaponToUI.new(self)
		end

		attr_reader :name
		attr_reader :uses
		attr_reader :type

		def power
			@type.power()
		end

		def useIt
			if uses > 0
				@uses -= 1
				power()
			else
				1.0
			end
		end
		
		def to_s
			return "Name: " + @name + ", Type: " + @type.to_s + ", Power: #{@type.power}, Uses: #{@uses}"
		end
		
	end	
end

File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/MARCIANITOS_P3/Marcianitos/src/deepspace/Weapon.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author mario
 */
class Weapon {
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon (String the_name, WeaponType the_type, int the_uses){
        this.name = the_name;
        this.type = the_type;
        this.uses = the_uses;
    }
    
    Weapon (Weapon w){
        this.name = w.name;
        this.type = w.type;
        this.uses = w.uses;
    }
    
    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }
    
    public WeaponType getType (){
        return (this.type);
    }
    
    public int getUses (){
        return (this.uses);
    }
    
    public float power (){
        return (type.getPower());
    }
    
    public float useIt(){
        float retorno = 1;
        
        if (this.uses > 0){
            --uses;
            retorno = this.power();
        }
        
        return (retorno);
    }
    
    public String toString(){
        
        String salida;
        
        salida = "Name=" + name + ", Type=" + type + ", Uses=" + uses;
        
        return salida;
    }
}
